DRAFT: MPI Fortran tutorial and example code

stephenpcook opened the pull request:
This is a draft version of a MPI Fortran tutorial.  I would appreciate feedback on the tutorial in `index.md`.

I think I need to run the example codes by someone more experienced with Fortran, and I still have some changes to make, such as
- [ ] Checking string formatting
- [ ] Make `my_bcast` take types other than integers
- [ ] Replacing relational operators, e.g. `.eq.` with `==`
- [x] Check best practice for including modules in a main file vs using `contains` and subfunctions.
- [ ] Add copyright notices to all code

Closes #99

stephenpcook committed:
Add Fortran 90 implementions of routines

stephenpcook committed:
Add Fortran implementations for broadcast code

stephenpcook committed:
Tidy up variable declarations

stephenpcook committed:
Add Fortran example of scatter, gather & allgather

stephenpcook committed:
Fortran examples for reduce and allreduce

stephenpcook committed:
Minor formatting

stephenpcook committed:
Fortran example of splitting communicator

stephenpcook committed:
Clean up declarations based on compiler warnings

Add warnings flag to makefile and remove missing dependency.

stephenpcook committed:
Fortran example of communicator groups

stephenpcook committed:
Rename fortran split and group communicators code

stephenpcook committed:
Update to Fortran 2008 MPI library

stephenpcook committed:
Bring fortran Hello World more in line with C code

stephenpcook committed:
Add MPI Fortran tutorial

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+    compute_avg = sum(array) / real(num_elements)
+  end function compute_avg
+end module subs
+
```
Not sure what the benefit of these functions are? Why not just inline the relevant code where it is needed? Also modules usually go in seperate files, and I've never seen it in the file containing the main program.

stephenpcook replied:
Module blocks removed in favour of external functions/subroutines and associated interface blocks.

I've also in-lined the random submodules, but I'm keeping the `compute_avg` for now to match the original C.

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+    integer :: i
+
+    do i = 1, num_elements
+      rand_nums(i) = rand()
```
The Fortran standard random function is called `random_number()`, I think `rand()` might be compiler specific (though probably ubiquitous).

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+    write (error_unit, *) 'Usage: all_avg num_elements_per_proc'
+    stop
+  end if
+
```
Nice defensive programming here.

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+  call get_command_argument(1, arg)
+
+  read (arg, *) num_elements_per_proc
+  ! Seed the random number generator to get different results each time
```
For defensive programming use the `IOSTAT` argument to the `read` function. Just in case the user passed a string of letters and you trying to read that into an integer.

stephenpcook replied:
Favouring matching the C version.

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+  read (arg, *) num_elements_per_proc
+  ! Seed the random number generator to get different results each time
+  call srand(time())
+  ! Throw away first rand value
```
I suggest using the Fortran standard `random_seed()` intrinsic instead. Calling it with no arguments will set the seed based on some random data retrieved form the operating system if I recall correctly.

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+  call srand(time())
+  ! Throw away first rand value
+  r = rand()
+
```
As suggested previously use the`random_number()` function instead.

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+  call MPI_Barrier(MPI_COMM_WORLD)
+  call MPI_FINALIZE()
+
+end program main
```
Since Fortran is case insensitive, I prefer using lower case for all variable names, function names and module names. Its a matter of taste I suppose, but here is a not official [Fortran style guide](https://fortran-lang.org/learn/best_practices/style_guide/)

stephenpcook replied:
I've seen MPI calls with the above case - this also matches the C functions, so I'm going to leave them as is.

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+end module subs
+
+program main
+  use mpi_f08
```
Not sure if this will bugger up the program, because of weird scoping side effects in the MPI module. But I normally always use the `only` specifier with `use` keyword. This will ensure no unintended side effects.

stephenpcook replied:
It seems to work fine, and there are already "rules" to not name any functions `MPI_*` so I'm going to leave this.

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+
+program main
+  use mpi_f08
+  use iso_fortran_env, only: error_unit
```
Good use of the `only` specifier.

ricky-lv426 commented on tutorials/mpi-fortran/code/avg.f90:
```diff
@@ -0,0 +1,105 @@
+
+    compute_avg = sum(array) / real(num_elements)
+  end function compute_avg
+end module subs
```
See my comment earlier of just using the relevant lines of code inline where it is needed instead of wrapping it in functions like this. The other option is to place this module in a separate file and compile all the individual programs against it.

stephenpcook replied:
I've moved this function to be a program-external one - not going to inline in for now, to match the C version.

ricky-lv426 commented on tutorials/mpi-fortran/code/avg.f90:
```diff
@@ -0,0 +1,105 @@
+end module subs
+
+program main
+  use mpi_f08
```
As suggested above use the `only` specifier and only import what you use.

ricky-lv426 commented on tutorials/mpi-fortran/code/avg.f90:
```diff
@@ -0,0 +1,105 @@
+  implicit none
+
+  integer :: num_args
+  character(12) :: arg
```
Just noticed this is a hard coded value of 12. Perhaps define it as a parameter? If you want to the code to be really flexible and help with ease of maintenance, create a seperate module file that contains default settings/values to create uniform behaviour accross all programs in this pakage. Then import (`use`) values as needed.

ricky-lv426 commented on tutorials/mpi-fortran/code/avg.f90:
```diff
@@ -0,0 +1,105 @@
+
+  call get_command_argument(1, arg)
+
+  read (arg, *) num_elements_per_proc
```
See my comment previously of using the `iostat` option in the `read()` function.

ricky-lv426 commented on tutorials/mpi-fortran/code/avg.f90:
```diff
@@ -0,0 +1,105 @@
+
+  read (arg, *) num_elements_per_proc
+  ! Seed the random number generator to get different results each time
+  call srand(time())
```
See previous comment on using the `random_seed()` function instead.

ricky-lv426 commented on tutorials/mpi-fortran/code/avg.f90:
```diff
@@ -0,0 +1,105 @@
+  ! Seed the random number generator to get different results each time
+  call srand(time())
+  ! Throw away first rand value
+  r = rand()
```
See previous comment on using the `random_number()` function instead.

ricky-lv426 commented on tutorials/mpi-fortran/code/all_avg.f90:
```diff
@@ -0,0 +1,98 @@
+  implicit none
+
+  integer :: num_args
+  character(12) :: arg
```
See comment later on about using a hardcoded value of 12 here.

ricky-lv426 commented on tutorials/mpi-fortran/code/check_status.f90:
```diff
@@ -0,0 +1,48 @@
+
+  if (world_rank .eq. 0) then
+    ! Pick a random amount of integers to send to process one
+    call srand(time())
```
See comment earlier about using `random_seed()` instead.

ricky-lv426 commented on tutorials/mpi-fortran/code/check_status.f90:
```diff
@@ -0,0 +1,48 @@
+    call srand(time())
+
+    ! Throw away first value
+    r = rand()
```
See comment earlier about using `random_number()` instead.

ricky-lv426 commented on tutorials/mpi-fortran/code/check_status.f90:
```diff
@@ -0,0 +1,48 @@
+
+  call MPI_FINALIZE()
+
+end program
```
Should there not be a check somewhere to ensure there is more than one rank  for this code to work? Though I am not sure a program compiled with MPI would actually try to run on one rank only!

stephenpcook replied:
The structure matches the C tutorials, so I'll leave as is.

ricky-lv426 commented on tutorials/mpi-fortran/code/comm_groups.f90:
```diff
@@ -0,0 +1,51 @@
+
+  call MPI_Finalize()
+
+end program main
```
Should there not be check somewhere that the total number of ranks is at least 14?

stephenpcook replied:
Matching the C version.

ricky-lv426 commented on tutorials/mpi-fortran/code/comm_split.f90:
```diff
@@ -0,0 +1,32 @@
+program main
+  use mpi_f08
```
See previous comments about using `only` specifier perhaps.

ricky-lv426 commented on tutorials/mpi-fortran/code/comm_groups.f90:
```diff
@@ -0,0 +1,51 @@
+program main
+  use mpi_f08
```
See previous comments about using `only` specifier perhaps.

ricky-lv426 commented on tutorials/mpi-fortran/code/compare_bcast.f90:
```diff
@@ -0,0 +1,96 @@
+    call MPI_RECV(data, count, datatype, root, 0, communicator, MPI_STATUS_IGNORE, ierror)
+  end if
+end subroutine my_bcast
+end module subs
```
Rather than putting a module inside the same file as the main program, instead make the subroutine `my_bcast()` a subprogram of the main one using `contains`?

stephenpcook replied:
Still in the same file, but improved layout.

ricky-lv426 commented on tutorials/mpi-fortran/code/compare_bcast.f90:
```diff
@@ -0,0 +1,96 @@
+end module subs
+
+program main
+  use mpi_f08
```
See previous comments about using `only` specifier perhaps.

ricky-lv426 commented on tutorials/mpi-fortran/code/compare_bcast.f90:
```diff
@@ -0,0 +1,96 @@
+  call get_command_argument(2, args(2))
+
+  read (args(1), *) num_elements
+  read (args(2), *) num_trials
```
See my previous comment on using `iostat` option to `read()` function to catch invalid input.

ricky-lv426 commented on tutorials/mpi-fortran/code/compare_bcast.f90:
```diff
@@ -0,0 +1,96 @@
+  implicit none
+
+  integer              :: num_args, num_elements, num_trials
+  character(12)        :: args(2)
```
As previously suggested maybe make the 12 a parameter that can be imported from a module.

ricky-lv426 commented on tutorials/mpi-fortran/code/compare_bcast.f90:
```diff
@@ -0,0 +1,96 @@
+    ! Synchronize again before obtaining final time
+    call MPI_Barrier(MPI_COMM_WORLD)
+    total_my_bcast_time = total_my_bcast_time + MPI_Wtime()
+
```
The logic in this block (line 70 to 75) works. However line 70  seems misleading. I'd rather define a variable on line 70:

`ref_wall_clock time =  mpi_wtime()`

and then replace line 75 with:

`total_my_bcast_time = total_my_bcast_time + (mpi_wtime() - ref_wall_clock time)`

stephenpcook replied:
This is syntax from the C version.

ricky-lv426 commented on tutorials/mpi-fortran/code/compare_bcast.f90:
```diff
@@ -0,0 +1,96 @@
+
+    ! Time MPI_Bcast
+    call MPI_Barrier(MPI_COMM_WORLD)
+    total_mpi_bcast_time = total_mpi_bcast_time - MPI_Wtime()
```
See comment above on the logic in lines 70 to 75 above, and apply it here as well.

ricky-lv426 commented on tutorials/mpi-fortran/code/makefile:
```diff
@@ -0,0 +1,53 @@
+	$(MPIF90) $(FFLAGS) comm_groups.f90 -o comm_groups
+
+clean:
+	rm -f $(EXECS) $(MOD_FILES)
```
I'm no makefile guru but this looks all good to me. Of course have to change the file slightly if any parameters are moved into a separate module.

stephenpcook replied:
Modules removed.

ricky-lv426 commented on tutorials/mpi-fortran/code/mpi_hello_world.f90:
```diff
@@ -0,0 +1,30 @@
+  ! Finalize the MPI environment
+  call MPI_FINALIZE()
+
+end program
```
Looks all good to me!

ricky-lv426 commented on tutorials/mpi-fortran/code/my_bcast.f90:
```diff
@@ -0,0 +1,56 @@
+module subs
+  use mpi_f08
+contains
+subroutine my_bcast(data, count, datatype, root, communicator, ierror)
```
Suggestion: Indent code here (and anywhere else after a `contains` by 2  further spaces)

ricky-lv426 commented on tutorials/mpi-fortran/code/my_bcast.f90:
```diff
@@ -0,0 +1,56 @@
+  ! Finalize the MPI environment
+  call MPI_FINALIZE()
+
+end program
```
Looks all fine, logic wise, to me. I would personally extract the subs module in a separate module file (and name it something else like `my_bcast_mod`).

stephenpcook replied:
Changed to an external subroutine in the same file - duplication with `compare_bcast.f90`, but the examples can run independently.

ricky-lv426 commented on tutorials/mpi-fortran/code/ping_pong.f90:
```diff
@@ -0,0 +1,39 @@
+  call MPI_COMM_RANK(MPI_COMM_WORLD, world_rank)
+
+  ping_pong_count = 0
+  ping_pong_limit = 10
```
Make `ping_pong_limit` a parameter rather than a variable.

ricky-lv426 commented on tutorials/mpi-fortran/code/ping_pong.f90:
```diff
@@ -0,0 +1,39 @@
+  ping_pong_count = 0
+  ping_pong_limit = 10
+
+  partner_rank = mod(world_rank + 1, 2)
```
Is this "game" only being played using two ranks (if so add a check to that effect)? If more than two ranks it looks like each odd rank is partnered with rank 0 and each even rank is partnered with rank 1?

stephenpcook replied:
Two ranks for this tutorial - no checks matches C version.

ricky-lv426 commented on tutorials/mpi-fortran/code/probe.f90:
```diff
@@ -0,0 +1,53 @@
+
+  if (world_rank .eq. 0) then
+    ! Pick a random amount of integers to send to process one
+    call srand(time())
```
See previous comment about using the Fortran standard `random_seed()` instead?

ricky-lv426 commented on tutorials/mpi-fortran/code/probe.f90:
```diff
@@ -0,0 +1,53 @@
+    call srand(time())
+
+    ! Throw away first value
+    r = rand()
```
See previous comment about using Fortran standard `random_number()` instead.

ricky-lv426 commented on tutorials/mpi-fortran/code/probe.f90:
```diff
@@ -0,0 +1,53 @@
+    ! Throw away first value
+    r = rand()
+
+    number_amount = int(rand() * real(MAX_NUMBERS))
```
Use Fortran standard `random_number()` instead?

ricky-lv426 commented on tutorials/mpi-fortran/code/reduce_avg.f90:
```diff
@@ -0,0 +1,78 @@
+    end do
+
+  end subroutine create_rand_nums
+end module subs
```
As per previous comments, what is the benefit of this module over just inlining the necessary code where it is needed?

ricky-lv426 commented on tutorials/mpi-fortran/code/reduce_avg.f90:
```diff
@@ -0,0 +1,78 @@
+  end if
+
+  call get_command_argument(1, arg)
+  read (arg, *) num_elements_per_proc
```
As per previous comment use `iostat` optional argumant to `read() `function to catch dodgy input.

ricky-lv426 commented on tutorials/mpi-fortran/code/reduce_avg.f90:
```diff
@@ -0,0 +1,78 @@
+
+  ! Create a random array of elements on all processes.
+  call srand(time()) ! Seed the random number generator to get different results each time for each processor
+  ! Throw away first rand value
```
Explain why?

stephenpcook replied:
After `srand` the first call to `rand` always returns a large number (it looks like the seed used), so I was throwing this away.  This issue has gone away when using `random_seed` and `random_number`.

ricky-lv426 commented on tutorials/mpi-fortran/code/reduce_std.f90:
```diff
@@ -0,0 +1,89 @@
+    end do
+
+  end subroutine create_rand_nums
+end module subs
```
As per previous comments what is the benefit of using this module when you can just inline the few required lines of code where it is needed?

ricky-lv426 commented on tutorials/mpi-fortran/code/reduce_std.f90:
```diff
@@ -0,0 +1,89 @@
+  implicit none
+
+  integer :: num_args
+  character(12) :: arg
```
As per previous suggestion make the 12 a parameter than can be imported from a module?

ricky-lv426 commented on tutorials/mpi-fortran/code/reduce_std.f90:
```diff
@@ -0,0 +1,89 @@
+  end if
+
+  call get_command_argument(1, arg)
+  read (arg, *) num_elements_per_proc
```
As per previous comment use `iostat` optional argument in `read()` function to checking for dodgy input.

ricky-lv426 commented on tutorials/mpi-fortran/code/reduce_std.f90:
```diff
@@ -0,0 +1,89 @@
+
+  ! Create a random array of elements on all processes.
+  call srand(time()) ! Seed the random number generator of processes uniquely
+  ! Throw away first rand value
```
Explain why?

ricky-lv426 commented on tutorials/mpi-fortran/code/ring.f90:
```diff
@@ -0,0 +1,39 @@
+
+  call MPI_FINALIZE()
+
+end program
```
Looks all fine to me.

ricky-lv426 commented on tutorials/mpi-fortran/code/send_recv.f90:
```diff
@@ -0,0 +1,27 @@
+  call MPI_COMM_RANK(MPI_COMM_WORLD, world_rank)
+
+  if (world_rank .eq. 0) then
+    num = -1
```
Turn this into a parameter?

ricky-lv426 reviewed (changes requested):
Hi Stephen it looks generally fine to me. Haven't seem MPI code for a while!

I have made a couple suggestions/recommendations mainly wrt the structure of the code. I think I caught all cases but basically they mostly fall into:

- Either remove unnecessary modularised code as there seems no benefit over just inlining the relevant code where it is needed. Or perhaps create an external module to import set parameters, which would e.g. make future changes to code input easier to modify.
- Use Fortran standard intrinsics.
- Do some more stringent checks on inputs.

Nice little tutorial you've put together. Hopefully my comments/suggestions are helpful.

stephenpcook committed:
Remove module blocks and refactor RNG calls

Module blocks removed in favor of external functions/subroutines and
interface blocks.

Random number functions `srand` and `rand` replaced by more standard
`random_seed` and `random_number` calls.

stephenpcook committed:
Update bcast examples with subroutines/interfaces

liamjberrisford commented:
# Pedagogical Review of `index.md`

## Define Clear Learning Objectives
- Introduce a brief list of learning objectives at the beginning of the page using action-oriented verbs (e.g. describe, implement, compare).
- Ensure the objectives are aligned with Bloom's Taxonomy, and build up and reflect both conceptual understanding and practical application.

## Improve Instructional Design and Structure
- Follow a coherent instructional design model such as ADDIE or Backwards Design. For example
  - Begin with foundational concepts (What is MPI?)
  - Progress to syntax and differences in Fortran
  - Follow with practical code examples
  - End with a summary, next steps, or further exercises
- Ensure each section supports the achievement of the stated objectives.
In its current form it sort of reads like documentation rather than an actual lesson on MPI?

## Reduce Cognitive Load
- Break large blocks of code or explanation into smaller, more manageable units.
- Insert explanations between code blocks to support gradual comprehension.
- Minimise the use of jargon or provide definitions and context for unfamiliar terms.
I think the place where this is particularly evident is within the first code block that is given. I understand the desire for a runnable Hello World example, but it feels like it jumps into a very big code block. too quickly

## Incorporate Inclusive and Accessible Design
- Add visual aids (e.g. diagrams or illustrations of MPI process flow) to support multiple learning styles. 

## Introduce Formative Assessment
- Add short reflective or diagnostic questions throughout the text to reinforce learning.
  - Example: "What is the role of MPI_COMM_WORLD in both the Fortran and C versions?"
- Depending on what you are using, there are lots of packages that provide interactive quizzes, such as JupyterQuiz, but other packages that create GitHub Pages content also have it. 

## Encourage a Growth Mindset and Normalise Errors
- Share examples of common mistakes and how to debug them effectively.

## Provide Real-World Context
- Briefly explain how MPI is used in practical scientific or industrial applications? I think its always handy to have some context for where they are going with the content and the justification for being able to use it.

stephenpcook committed:
Add MPI-Fortran to tutorials index
